20220714-lookup-api.md

```diff
@@ -0,0 +1,457 @@
+
+We should modify the [`TupleBackend`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L47) interface with a new method `QueryRelationshipTuples` and it's signature could look like:
+
+```
+QueryRelationships(
+  ctx context.Context,
+  storeID string,
+  filter *RelationshipFilter,
+  opts ...QueryOptionsOption,
+) (TupleIterator, error)
+
+type RelationshipFilter struct {
+  ObjectType         string
+  OptionalObjectID   string
+  OptionalRelation   string
+  OptionalUserFilter string
+}
+
+type QueryOptions struct {
+  Limit    *uint64
+  Usersets []string
+}
+
+type QueryOptionsOption func(*QueryOptions)
+```
+
+This approach will allow us to query relationship tuples using any provided filters, and could serve as a more general replacement for the existing Read method.
+
+# Migration
+[migration]: #migration
+
+No migrations or breaking changes are necessary to support the Lookup API endpoints described herein.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Potential arguments for not introducing the Lookup API include:
+* Without proper limitations clients could impact the whole system by issuing Lookups on large collections.
+* Increases the OpenFGA API surface for a very niche use case for Search with Permissions on small object collections.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## More Generic Lookup API
+An alternative version of Lookup could support both generic and specific queries such as:
+> list all documents that user:bob can 'read' or 'write'
+
+For example,
+```
+curl --request POST 'http://localhost:8080/stores/<storeID>/lookup' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "object_type": "document",
+    "relations": ["read", "write"],
+    "user": "bob",
+    "authorization_model_id": "<modelID>"
+}'
+```
+and
+
+> list all relationships for documents that user:bob has
+
+For example,
+```
+curl --request POST 'http://localhost:8080/stores/<storeID>/lookup' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "object_type": "document",
+    "user": "bob",
+    "authorization_model_id": "<modelID>"
+}'
+```
+
+This alternative Lookup API definition is more generic and could be more generally applicable to specific cases where you want to know all of the objects (of a specific type) and the relationships to those objects a user has. However, the query cost of supporing a generic API endpoint like this could be quite expensive since you may have to evaluate all object and relation pairs for objects of a given type.
```

Comment:
What would be the use-case for this approach?

Reply:
If you're pre-filtering a collection of objects based on one or more relationships. For example, if you wanted to know the objects the user has 'edit' and 'delete' access to so that you can show a pencil and trash can icon in a UI.

Reply:
Should we make the API response to be more generic to allow this case in the future if we see a reason for it?

Response would become:
- user: ..., relation: ..., object: ...

In the future if you'd need to expand support, the interface would not need to change